bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeChecker.java

```diff
@@ -1040,18 +1045,83 @@ private static boolean checkIsArrayType(BType sourceType, BArrayType targetType,
     }
 
     private static boolean checkIsTupleType(BType sourceType, BTupleType targetType, List<TypePair> unresolvedTypes) {
-        if (sourceType.getTag() != TypeTags.TUPLE_TAG) {
+
+        if (sourceType.getTag() == TypeTags.UNION_TAG) {
+            for (BType memberType : ((BUnionType) sourceType).getMemberTypes()) {
+                if (!checkIsTupleType(memberType, targetType, unresolvedTypes)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        if (sourceType.getTag() != TypeTags.ARRAY_TAG && sourceType.getTag() != TypeTags.TUPLE_TAG) {
+            return false;
+        }
+
+        List<BType> targetTypes = new ArrayList<>(targetType.getTupleTypes());
+        BType targetRestType = targetType.getRestType();
+
+        BTupleType sourceTupleType;
+        if (sourceType.getTag() == TypeTags.TUPLE_TAG) {
+            sourceTupleType = (BTupleType) sourceType;
+        } else {
+            BArrayType sourceArrayType = (BArrayType) sourceType;
+            BType sourceElementType = sourceArrayType.getElementType();
+
+            switch (sourceArrayType.getState()) {
+                case UNSEALED:
+                    if (targetRestType == null) {
+                        return false;
+                    } else {
+                        return checkIsType (sourceElementType, targetRestType, unresolvedTypes);
+                    }
+                case CLOSED_SEALED:
+                    if (sourceArrayType.getSize() >= targetTypes.size()) {
+                        if (targetTypes.isEmpty()) {
+                            if (targetRestType != null) {
+                                return checkIsType (sourceElementType, targetRestType, unresolvedTypes);
+                            }
+                            return true;
```

Comment:
Shouldn't we check the size here? Reaching here would mean the target is an empty list? This would mean the source also has to be empty?

Reply:
Actually I was wrong here. Can a tuple  exist with its member type size being empty and also the rest type being null?

Reply:
Although we don't seem to allow writing such a tuple atm, I believe it is valid according to the spec. IMO, we can add the validation now since it is a simple check. That way it will be future-proof too.